src/main/java/engineering/swat/watch/impl/mac/NativeEventStream.java

```diff
@@ -0,0 +1,286 @@
+import engineering.swat.watch.impl.mac.apis.DispatchQueue;
+import engineering.swat.watch.impl.mac.apis.FileSystemEvents;
+import engineering.swat.watch.impl.mac.apis.FileSystemEvents.FSEventStreamCallback;
+
+// Note: This file is designed to be the only place in this package where JNA is
+// used and/or the native APIs are invoked. If the need to do so arises outside
+// this file, consider extending this file to offer the required services
+// without exposing JNA and/or the native APIs.
+
+/**
+ * <p>
+ * Stream of native events for a path, issued by macOS. It's a facade-like
+ * object that hides the low-level native APIs behind a higher-level interface.
+ * </p>
+ *
+ * <p>
+ * Note: Methods {@link #open()} and {@link #close()} synchronize on this object
+ * to avoid races. The synchronization overhead is expected to be negligible, as
+ * these methods are expected to be rarely invoked.
+ * </p>
+ */
+public class NativeEventStream implements Closeable {
+
+    // Native APIs
+    private static final CoreFoundation   CF  = CoreFoundation.INSTANCE;
+    private static final DispatchObjects  DO  = DispatchObjects.INSTANCE;
+    private static final DispatchQueue    DQ  = DispatchQueue.INSTANCE;
+    private static final FileSystemEvents FSE = FileSystemEvents.INSTANCE;
+
+    // Native memory (automatically deallocated when set to `null`)
+    private @Nullable FSEventStreamCallback callback;
+    private @Nullable Pointer stream;
+    private @Nullable Pointer queue;
+
+    private final Path path;
+    private final NativeEventHandler handler;
+    private volatile boolean closed;
+
+    public NativeEventStream(Path path, NativeEventHandler handler) throws IOException {
+        this.path = path.toRealPath(); // Resolve symbolic links
+        this.handler = handler;
+        this.closed = true;
+    }
+
+    public synchronized void open() {
+        if (!closed) {
+            throw new IllegalStateException("Stream is already open");
+        } else {
+            closed = false;
+        }
+
+        // Allocate native memory. (Checker Framework: The local variables are
+        // `@NonNull` copies of the `@Nullable` fields.)
+        var callback = this.callback = createCallback();
+        var stream = this.stream = createFSEventStream(callback);
+        var queue = this.queue = createDispatchQueue();
+
+        // Start the stream
+        FSE.FSEventStreamSetDispatchQueue(stream, queue);
+        FSE.FSEventStreamStart(stream);
+    }
+
+    private FSEventStreamCallback createCallback() {
+        return new FSEventStreamCallback() {
+            @Override
+            public void callback(Pointer streamRef, Pointer clientCallBackInfo,
+                    long numEvents, Pointer eventPaths, Pointer eventFlags, Pointer eventIds) {
+                // This function is called each time native events are issued by
+                // macOS. The purpose of this function is to perform the minimal
```

Comment:
I might be wrong, but it reads like we should make sure that this function does not consume a lot of time? As it's a native callback?

If that is true, I think we should publish the event to a `ConcurrentQueue`, as we don't know how much work happens inside of the `handler`